### PR TITLE
Add headings with IDs plugin

### PR DESCRIPTION
**Pull request by lucperkins**
I've been enjoying this library a bunch recently so I thought I'd contribute an example as thanks. I'm not sure how idiomatic this implementation is but I think it provides a decent real-world use case.

### CODE REVIEW

**Comment by rlidwka** [comment]
Thanks, that is indeed a real-world use case.

Should it be an example, or just a plugin (in extras, like syntax highlighting is now)? For markdown-it we have https://github.com/valeriangalliat/markdown-it-anchor, maybe it's useful to add a plugin with similar api.

**Comment by lucperkins** [comment]
@rlidwka I'm certainly happy to make it a plugin would be happy to update the API to be a bit more robust à la `markdown-it-anchor`!

File: src/plugins/extra/heading_anchors.rs [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+        inline::{
+            backticks::CodeInline,
+            emphasis::{Em, Strong},
+            link::Link,
+        },
+    },
+    MarkdownIt, Node, NodeValue, Renderer,
+};
+
+use super::strikethrough::Strikethrough;
```
Comment (rlidwka) [inline src/plugins/extra/heading_anchors.rs:16]: inconsistent import style, this uses `super::`, and there are `crate::` imports above

File: src/plugins/extra/heading_anchors.rs [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+    }
+}
+
+struct HeadingsWithIds;
+
+impl CoreRule for HeadingsWithIds {
+    fn run(node: &mut Node, _: &MarkdownIt) {
+        // Walk the existing AST and replace all headings with our custom headings
+        node.walk_mut(|node, _| {
+            if let Some(heading) = node.cast::<ATXHeading>() {
```
Comment (rlidwka) [inline src/plugins/extra/heading_anchors.rs:46]: Setext heading is missing.

There are two styles in markdown:

```md
## atx heading

setext heading
==============
```

This rule only covers first, but I guess both needs to be present?

It was my questionable design decision to make separate nodes for both types of headings. If we are to follow it, this code has to be duplicated with `ATXHeadingWithSlug` and `SetextHeadingWithSlug`. *(Or should I make a large change and unify those tokens?)*

File: src/plugins/extra/heading_anchors.rs [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+// `Here is **strong** text` needs to be coverted to the string `Here is strong
+// text` (and then turned into the ID `here-is-strong-text`). There's probably
+// a more elegant way to do this but this works for me.
+fn inline_node_to_string(node: &Node) -> String {
+    let mut pieces: Vec<String> = Vec::new();
+
+    for node in node.children.iter() {
+        if let Some(txt) = node.cast::<Text>() {
+            pieces.push(txt.content.clone());
+        } else if node.is::<CodeInline>()
```
Comment (rlidwka) [inline src/plugins/extra/heading_anchors.rs:69]: There is more elegant way - traverse the tree and collect Text nodes:

https://github.com/rlidwka/markdown-it.rs/blob/eb5459039685d19cefd0361859422118d08d35d4/src/plugins/cmark/inline/image.rs#L21-L29

**Review by rlidwka — COMMENTED** [review]
sorry for the delay, looks good to me

-----

Is there any useful configuration for this plugin?

As an idea, we could pass slugify function as config option (either `fn (&str) -> String` or boxed closure). This way we don't need to depend on slugify, and users will have freedom to customize slugify function.

**Comment by ShafSpecs** [comment]
@lucperkins any update to this PR? As mentioned, it's a great real life implementation that would be nice to have

**Comment by rlidwka** [comment]
I've implemented it from scratch in #18, it's actually much simpler than proposed here (e.g. no new tags are required, you can push attributes into existing ones).

Not completely sure we need it, since you can copy-paste it into your own project. This is the entire thing:

```rs
use markdown_it::parser::core::CoreRule;
use markdown_it::plugins::cmark::block::heading::ATXHeading;
use markdown_it::plugins::cmark::block::lheading::SetextHeader;
use markdown_it::{MarkdownIt, Node};

fn main() {
    let md = &mut MarkdownIt::new();
    markdown_it::plugins::cmark::add(md);

    md.add_rule::<AddHeadingAnchors>();

    pub struct AddHeadingAnchors;
    impl CoreRule for AddHeadingAnchors {
        fn run(root: &mut Node, _md: &MarkdownIt) {
            root.walk_mut(|node, _| {
                if node.is::<ATXHeading>() || node.is::<SetextHeader>() {
                    // collect_text() is new api proposed in #18
                    node.attrs.push(("id", slug::slugify(node.collect_text())));
                }
            });
        }
    }

    dbg!(md.parse("## An example heading").render());
}
```

**Comment by rlidwka** [comment]
published as 0.5.0, closing as resolved

**Pull request closed without merging**